### DISCUSSION TITLE
Mistake in Supabase doc ?

### DISCUSSION THREAD

**Discussion by Lu-Family** [supabase/supabase#27619] [Questions]
Hi, 

I'm trying to implement a supabase auth with a next.js app.  

I'm following this: https://supabase.com/docs/guides/auth/server-side/nextjs 

But something seems wrong in the middleware part. 
When I retrieve my user `const { data: user, error } = await supabase.auth.getUser()` 

It seems like the user object returned from supabase.auth.getUser() is structured in a way that user itself is an object containing null. This structure suggests that even though user is not null in a strict sense, its user property inside it is null.

 `if (!user) {
    // no user, potentially respond by redirecting the user to the login page
    return NextResponse.redirect('/login')
  }` 
It looks like I should  adjust the conditional check to access user.user instead of just user 

And looking at this: https://nextjs.org/docs/messages/middleware-relative-urls 
it says: "Prior to Next.js 12.1, we allowed passing relative URLs. However, constructing a request with new Request(url) or running fetch(url) when url is a relative URL does not work. " 

So the redirect is not working either. 

Am I missing something ? Is the doc outdated ? 

Thanks

**Comment by ErikPetersenDev** [comment]
I ran into these documentation issues as well when updating `supabase/ssr` to 0.4.0.

For the first issue, they are missing the brackets around "user" to destructure it.  Changing it to `data: { user }` fixes that issue.

And you're correct about the middleware as well.  Once the `user` property is accessed correctly the existing code throws an error, as discussed in the Next.js doc you linked.
Instead of `NextResponse.redirect('/login')`
the code should be (recommended solution from the Next.js doc):
```
    const url = request.nextUrl.clone();
    url.pathname = "/login";
    return NextResponse.redirect(url);
```

There's also a comment at the bottom of the same file (`utils/supabase/middleware.ts`) that indicates you should add the line `myNewResponse.cookies.setAll(supabaseResponse.cookies.getAll())` in a certain situation.  According to an issue someone else filed, `setAll` doesn't exist in this scenario.  Issue: https://github.com/supabase/supabase/issues/27505

Rather than the docs being outdated, it looks like these issues were introduced recently when the docs were modified to reflect `@supabase/ssr` changes in 0.4.0.  Here's the PR: https://github.com/supabase/supabase/pull/27242.

EDIT:   I forgot to mention that you also need to make this change to the `if`:
`if (!user && !request.nextUrl.pathname.startsWith("/login")) {`

Without that change you'll end up in an infinite redirect loop!

Hope this helps!

**Reply by ErikPetersenDev** [reply]
I've opened a PR to address the issues discussed above:  https://github.com/supabase/supabase/pull/27622
